src/GraphQL/Types/IProvideMetadata.cs

```diff
@@ -5,36 +5,7 @@ namespace GraphQL.Types
     /// This interface is implemented by numerous descendants like <see cref="GraphType"/>,
     /// <see cref="FieldType"/>, <see cref="Schema"/> or others.
     /// </summary>
-    public interface IProvideMetadata
+    public interface IProvideMetadata : IMetadataBuilder
```

Comment:
From `IMetadataBuilder`:
> This interface is implemented by field builders for configuring metadata.

So each extension method that accepts `IMetadataBuilder` can accept `IProvideMetadata` as well. I do not see advantages in addition of second interface from the caller point of view. I understand what this PR about but jumping around changes `IProvideMetadata` <->`IMetadataBuilder` looks unnecessary.

Reply:
We certainly do not want to add methods such as IsAnonymous or GetDirective to field builders

Reply:
[GetDirective -> GetAppliedDirectives ?] Well, now I better understand intention but still - why not? We do not **add** all that methods. They just exist. Honestly, it's harder for me to read code with two nearly identical interfaces for different purposes. OK, if you **really** want to split usage of builders from usage of "pure" `IProvideMetadata` let's consider such design:
```cs
public interface IBuilder { } // nothing here for now, just marker interface for where constraint 
public class AnyBuilder<T, U, whatever> : IBuilder, .... { ... } // add IBuilder for every builder class in repo

// AllowAnonymous - example of usage
public static TMetadataProvider AllowAnonymous<TMetadataProvider>(this TMetadataProvider provider)
            where TMetadataProvider : IProvideMetadata
            where TMetadataProvider : IBuilder // new constraint
```

so builder-releated methods are available only for classes marked with `IBuilder` interface.

Reply:
No, it should not be simply builder classes but any classes that allow writing metadata.  Consider:

```cs
var field = new FieldType(...);
field.AllowAnonymous();
```

It would not be logical for `FieldType` to be marked with `IBuilder`.  At least not at this time.  Also, it's the reader methods that need an additional constraint, not the writing methods.  Otherwise we would end up with the same problem with the builder methods.

Reply:
> It would not be logical for `FieldType` to be marked with `IBuilder`

Yes. I didn't want that.


So the whole intent is to split read APIs (informational) from write APIs (builders)?

Reply:
Yes, or more specificially, separate read/write from write-only.  GraphQL.NET doesn't have read-only classes as a general rule, like for example `FieldType` is read/write -- I always wish we would have an `IFieldType` that's read only.

So the extension methods for `IMetadataBuilder` is for writing only whereas `IProvideMetadata` is read/write.

Reply:
If you remember, I spent quite a lot of time attempting to make `ISchema` and all descendants read-only, so you 'build' the schema and then it is read-only for its lifetime.  This didn't work out as the design changes necessary were so complex that the concept was nixed.  If it were so, we would have read-only extension methods and write-only extension methods.  Some classes may be read-write.  But at the moment the infrastructure is either write-only (builder classes) or read/write (everything else).

Reply:
> So the extension methods for `IMetadataBuilder` is for writing only whereas `IProvideMetadata` is read/write.

OK, now I completely understand. Give me time to think about it, I'm still in doubt. It looks unusual.

Reply:
We can have `IMetadataReader` and `IMetadataWriter` which both inherit from `IProvideMetadata`.

Reply:
I mean that we begin to artificially separate APIs by read/write applicability and this can harm when both required.

Reply:
I added `MetadataReader` property to pull reader from writer interface.

Reply:
ok